laurent22/joplin
# Desktop: Change Joplin Cloud login process to allow MFA via browser

This is a working-in-progress PR for the new login flow that we are going to use for Joplin Cloud.

This change is required to make the 2FA work in the Desktop.

With this PR, when the user tries to synchronise to Joplin Cloud we will show a new page that will instruct the user on how to authorise the application over Joplin Cloud. This process will involve the creation of a unique value that Joplin Cloud will use to identify the application that allowed the use of the recently created credentials.

<details><summary>Images</summary>
<p>

![Screenshot from 2024-01-14 00-34-49](https://github.com/laurent22/joplin/assets/5051088/db4ac732-bce7-42ca-bf52-a94370ba2add)
![Screenshot from 2024-01-14 00-35-32](https://github.com/laurent22/joplin/assets/5051088/fe209116-06db-40f7-b80c-c7d14905cf7b)
![Screenshot from 2024-01-14 00-35-46](https://github.com/laurent22/joplin/assets/5051088/7bb82b1a-3527-4c2f-814e-d7d42e114c61)
![Screenshot from 2024-01-14 00-35-06](https://github.com/laurent22/joplin/assets/5051088/3c7811f1-94b8-446d-88cf-1f6315fa893c)

</p>
</details> 

There are some missing things for now:
- [x] Generate a proper `unique_login_code`
- [x] Test if I didn't break anything for Dropbox or OneDrive
- [x] Test login without 2FA (it shouldn't interfere, just a sanity check)
- [x] Include information about the client in the URL for login (platform and type)
- [x] Bug: when clicking in synchronize I'm creating 4 session records in db.
  - [x] Check if this wasn't an issue before.
- [x] Review code from syncWizard/Dialog.tsx, there are things there aren't necessary anymore.
- [x] Make code stateless and move to library to be used in CLI and Mobile.
  - Not sure exactly how much I could move to a library, some parts will be similar like the request the request to `api/applications` with an interval, but a lot of code here is platform-specific.
- [x] Update platform, type and version on new requests to `/api/sessions`
  - We probably just need to update only the version value since the other won't change
  - Need to discuss it how do it https://github.com/laurent22/joplin/pull/9445#issuecomment-1867906111

### Testing

I still need to add automated tests, maybe I will need to mock so much stuff that won't really be very useful?

Some manual tests that are good to be done:
- Make the proper process, authorize, see synchronisation working.
- Open JoplinCloudLogin page two times and copy the URL, the unique login code should not be unique, not the same
- The page should only start making requests to `api/applications` after one of the buttons is clicked
- The page should not start a new interval of requests if one is already active
- Closing the page should clear the interval of requests
- Clicking the Synchronise button in the Sidebar when the SyncTarget is set to Joplin Cloud but the login process wasn't completed should open the JopliCloudLogin page

#### Testing upgrading version without 2FA
- Start an old version of Desktop
- Synchronise to Joplin Cloud
- Upgrade Desktop version
- Check if synchronisation still works (it should, but won't be using application credentials yet, only when user redo the login process)

#### Testing upgrading version and enabling 2FA
- Start an old version of Desktop
- Synchronise to Joplin Cloud
-  Enable 2FA in Joplin Cloud website
- Upgrade Desktop version
- Synchronisation should not be working until user go through the new login process.
